Content/Items/Magnet/Weapons.ThunderBeads.cs

```diff
@@ -0,0 +1,280 @@
+
+		public bool ableToHit = false;
+		public bool leftClick = true;
+
+		public float fade = 0.1f;
+
+		public Color baseColor = new(200, 230, 255);
+		public Color endColor = Color.Purple;
+
+		private Trail trail;
+		private Trail trail2;
+		private List<Vector2> cache;
+		private List<Vector2> cache2;
+
+		public Player Owner => Main.player[Projectile.owner];
+
+		public override string Texture => AssetDirectory.MagnetItem + Name;
+
+		public ThunderBeads_Whip() : base("Thunder Beads", 15, 0.87f, Color.Transparent)
+		{
+			xFrames = 1;
+			yFrames = 5;
+		}
+
+		public override void Load()
+		{
+			On.Terraria.Projectile.FillWhipControlPoints += OverrideWhipControlPoints;
+			base.Load();
+		}
+
+		public override int SegmentVariant(int segment)
+		{
+			return 1;
+		}
+
+		public override void SafeSetDefaults()
+		{
+			Projectile.localNPCHitCooldown = 1;
+		}
+
+		public override bool PreAI()
+		{
+			ManageCache();
+			if (!Main.dedServ)
+				ManageTrails();
+
+			for (int i = 0; i < cache.Count - 1; i++)
+			{
+				if (i % 3 != 0)
+					continue;
+
+				Lighting.AddLight(cache[i], Color.Cyan.ToVector3() * fade * 0.5f);
+			}
+
+			if (fade > 0.1f)
+				fade -= 0.05f;
+
+			if (embedded)
+			{
+				if (!leftClick && Main.mouseLeft)
+				{
+					ableToHit = true;
+					leftClick = true;
+				}
+
+				if (!Main.mouseLeft)
+					leftClick = false;
+
+				flyTime = Owner.itemAnimationMax * Projectile.MaxUpdates;
+				Projectile.rotation = Projectile.velocity.ToRotation() + MathHelper.PiOver2;
+				Projectile.spriteDirection = (!(Vector2.Dot(Projectile.velocity, Vector2.UnitX) < 0f)) ? 1 : -1;
+
+				Owner.heldProj = Projectile.whoAmI;
+				Owner.itemAnimation = Owner.itemAnimationMax - (int)(Projectile.ai[0] / Projectile.MaxUpdates);
+				Owner.itemTime = Owner.itemAnimation;
+
+				embedTimer--;
+				if (embedTimer < 0 || !target.active)
+				{
+					Projectile.friendly = false;
+					embedded = false;
+					return false;
+				}
+
+				Projectile.Center = target.Center;
+				Projectile.WhipPointsForCollision.Clear();
+				SetPoints(Projectile.WhipPointsForCollision);
+				return false;
+			}
+			return base.PreAI();
+		}
+
+		public override void ArcAI()
+		{
+			xFrame = 0;
+		}
+
+		public override bool ShouldDrawSegment(int segment)
+		{
+			return segment % 3 == 0;
+		}
+
+		public override void OnHitNPC(NPC target, int damage, float knockback, bool crit)
+		{
+			ableToHit = false;
+			if (!embedded)
+			{
+				Projectile.ownerHitCheck = false;
+				Projectile.damage /= 2;
+				this.target = target;
+				embedded = true;
+			}
+			else
+			{
+				Helper.PlayPitched("Magic/LightningShortest1", 0.5f, Main.rand.NextFloat(0f, 0.2f), target.Center);
+				fade = 1;
+				for (int i = 2; i < cache.Count - 2; i++)
+				{
+					Dust.NewDustPerfect(cache[i] + Main.rand.NextVector2Circular(6,6), ModContent.DustType<GlowLineFast>(), cache[i].DirectionTo(Owner.Center).RotatedByRandom(0.2f) * Main.rand.NextFloat(1, 2), 0, Color.Cyan, 0.4f);
```

Comment:
Might be worth adding a linebreak here or pull out a local for the velocity

Reply:
Fixed